### PR TITLE
MET-3126 Implement new Enrichment API

### PR DESCRIPTION
**Pull request by JoanaCMS**

### CODE REVIEW

File: metis-enrichment/metis-enrichment-rest/src/test/java/eu/europeana/enrichment/rest/OldEnrichmentControllerTest.java [hunk 1]
```diff
@@ -30,7 +30,7 @@
 import org.springframework.test.web.servlet.MockMvc;
 import org.springframework.test.web.servlet.setup.MockMvcBuilders;
 
-public class EnrichmentControllerTest {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-rest/src/test/java/eu/europeana/enrichment/rest/OldEnrichmentControllerTest.java:33]: Would it be easy to copy this test class and adapt it to the new enrichment controller? So that that class is covered by tests as well?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java [hunk 2]
```diff
@@ -52,12 +53,12 @@ private EnrichmentUtils() {
...
-  public static List<InputValue> extractValuesForEnrichmentFromRDF(RDF rdf) {
+  public static List<Pair<SearchValue, EnrichmentFields>> extractValuesForEnrichmentFromRDF(RDF rdf) {
     final ProxyType providerProxy = RdfProxyUtils.getProviderProxy(rdf);
-    final List<InputValue> valuesForEnrichment = new ArrayList<>();
+    final List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment = new ArrayList<>();
     for (EnrichmentFields field : EnrichmentFields.values()) {
-      List<InputValue> values = field.extractFieldValuesForEnrichment(providerProxy);
-      valuesForEnrichment.addAll(values);
+      List<SearchValue> values = field.extractFieldValuesForEnrichment(providerProxy);
+      values.stream().map(val -> valuesForEnrichment.add(new Pair<>(val, field)));
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java:61]: Use forEach here.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java [hunk 3]
```diff
@@ -64,6 +65,7 @@
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.math3.util.Pair;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:68]: More unused import statements.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 4]
```diff
@@ -38,27 +40,35 @@ public EnricherImpl(EntityMergeEngine entityMergeEngine,
...
-      entityMergeEngine.mergeEntities(rdf, entities, referencesForEnrichment);
+    if (valueEnrichmentInformation != null) {
+      final Set<EnrichmentFields> enrichmentFieldsSet = valueEnrichmentInformation.stream()
+          .map(Pair::getSecond).collect(Collectors.toSet());
+      final List<EnrichmentResultBaseWrapper> enrichmentResultBaseWrapperList = valueEnrichmentInformation
+          .stream().map(Pair::getFirst).collect(Collectors.toList());
+
+      if (CollectionUtils.isNotEmpty(enrichmentResultBaseWrapperList)) {
+        entityMergeEngine
+            .mergeEntities(rdf, enrichmentResultBaseWrapperList, enrichmentFieldsSet);
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:65]: I think there is a problem here. By calling it like this, you add all results and all fields, and you lose the mapping. Instead you have to loop through the pairs and call the mergeEntity method multiple times to make sure that results are only added to the fields that they belong to.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java [hunk 5]
```diff
@@ -409,19 +410,20 @@ public void mergeEntities(RDF rdf, List<EnrichmentBaseWrapper> enrichmentBaseWra
    * @param proxyLinkTypes Lookup of the link types to create in the europeana proxy. The keys are
    * the about values of the entities to add.
    */
-  public void mergeEntities(RDF rdf, List<EnrichmentBase> contextualEntities,
+  public void mergeEntities(RDF rdf, List<List<EnrichmentBase>> contextualEntities,
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:413]: Probably nicer to make this a list of entries, not a nested list, and to do a flatMap in the calling method. Then I think you can revert the changes in this method.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java [hunk 6]
```diff
@@ -409,19 +410,20 @@ public void mergeEntities(RDF rdf, List<EnrichmentBaseWrapper> enrichmentBaseWra
...
       final Set<EnrichmentFields> fields = links.stream().map(proxyLinkTypes::get)
           .filter(Objects::nonNull).flatMap(Set::stream).collect(Collectors.toSet());
-      convertAndAddEntity(rdf, entity, fields);
+      entitiesList.forEach(entity -> convertAndAddEntity(rdf, entity, fields));
     }
   }
 
 
-  private static Set<String> getSameAsLinks(EnrichmentBase contextualClass) {
+  private Set<String> getSameAsLinks(EnrichmentBase contextualClass) {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:426]: Why not leave this static?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/Dereferencer.java [hunk 7]
```diff
@@ -22,7 +23,7 @@
    * @return A list of RDF field names with the information associated with it
    * @throws DereferenceException
    */
-  List<EnrichmentBaseWrapper> dereferenceEntities(Set<String> resourceIds) throws DereferenceException;
+  List<EnrichmentResultBaseWrapper> dereferenceEntities(Set<String> resourceIds) throws DereferenceException;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/Dereferencer.java:26]: Just changed

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java [hunk 8]
```diff
@@ -15,6 +16,7 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.Function;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java:19]: Didn't notice, will remove them

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java [hunk 9]
```diff
@@ -66,10 +68,14 @@ public void dereference(RDF rdf) throws DereferenceException {
...
-        .map(EnrichmentBaseWrapper::getEnrichmentBase).map(EnrichmentBase::getAbout)
-        .collect(Collectors.toSet());
+    final List<EnrichmentResultBaseWrapper> result = new ArrayList<>(dereferenceOwnEntities(resourceIds));
+    final List<List<EnrichmentBase>> listOfListEnrichmentBase = result.stream()
+        .map(EnrichmentResultBaseWrapper::getEnrichmentBaseList)
+        .collect(Collectors.toList());
+
+    final Set<String> foundOwnEntityIds = listOfListEnrichmentBase.stream()
+        .map(this::getAllAboutField)
+        .flatMap(List::stream).collect(Collectors.toSet());
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java:78]: Removed the new method and made the statement simpler

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java [hunk 10]
```diff
@@ -82,8 +88,12 @@ public void dereference(RDF rdf) throws DereferenceException {
     return result;
   }
 
+  private List<String> getAllAboutField(List<EnrichmentBase> listEnrichmentBase){
+    return listEnrichmentBase.stream().map(EnrichmentBase::getAbout).collect(Collectors.toList());
+  }
+
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java:94]: Removed

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java [hunk 11]
```diff
@@ -2,13 +2,16 @@
...
+import eu.europeana.enrichment.api.external.model.EnrichmentResultBaseWrapper;
 import eu.europeana.enrichment.api.external.model.EnrichmentResultList;
 import eu.europeana.enrichment.rest.client.exceptions.EnrichmentException;
 import eu.europeana.enrichment.utils.EnrichmentFields;
 import eu.europeana.enrichment.utils.InputValue;
+import eu.europeana.enrichment.utils.SearchValue;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import org.apache.commons.math3.util.Pair;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java:14]: Didn't notice, will remove them

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java [hunk 12]
```diff
@@ -25,7 +28,7 @@
    * @return A enrichment result list with the information retrieved to enrich the RDF
    * @throws EnrichmentException
    */
-  EnrichmentResultList enrichValues(List<InputValue> valuesForEnrichment)
+  List<Pair<EnrichmentResultBaseWrapper, EnrichmentFields>> enrichValues(List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment)
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java:31]: I just changed it, will need to fix the unit tests afterwards

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java [hunk 13]
```diff
@@ -34,7 +37,7 @@ EnrichmentResultList enrichValues(List<InputValue> valuesForEnrichment)
    * @return A list with RDF fields retrieved and the information associated with it
    * @throws EnrichmentException
    */
-  List<EnrichmentBaseWrapper> enrichReferences(Set<String> referencesForEnrichment)
+  List<EnrichmentResultBaseWrapper> enrichReferences(Set<String> referencesForEnrichment)
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/Enricher.java:40]: Just changed it. I also had to change the return type of other methods

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java (hunk 4 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:65]: Should I change the signature of mergeEntities then? To Accept EnrichmentBase and EnrichmentFields alone? Or should I be careful with that inside mergeEntities?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 14]
```diff
@@ -71,20 +81,30 @@ public void enrichment(RDF rdf) throws EnrichmentException {
   }
 
   @Override
-  public EnrichmentResultList enrichValues(List<InputValue> valuesForEnrichment)
+  //TODO: Input parameter is List<Pair<SearchValue, EnrichmentFields>
+  //TODO: Output is List<Pair<EnrichmentResult, EnrichmentFields>>
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:85]: Yes I can, will do that

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnrichmentClient.java [hunk 15]
```diff
@@ -2,14 +2,20 @@
...
 
 import eu.europeana.enrichment.api.exceptions.UnknownException;
+import eu.europeana.enrichment.api.external.EnrichmentSearch;
 import eu.europeana.enrichment.api.external.InputValueList;
 import eu.europeana.enrichment.api.external.model.EnrichmentBase;
 import eu.europeana.enrichment.api.external.model.EnrichmentBaseWrapper;
+import eu.europeana.enrichment.api.external.model.EnrichmentResultBaseWrapper;
 import eu.europeana.enrichment.api.external.model.EnrichmentResultList;
 import eu.europeana.enrichment.utils.InputValue;
+import eu.europeana.enrichment.utils.SearchValue;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnrichmentClient.java:18]: Removed

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java (hunk 3 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:68]: Removed

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java (hunk 4 above)
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:65]: No, I think you should be careful with that here in this method. 

I think this should be wrapped in a loop over all pairs in `valueEnrichmentInformation` and it should be called once for each pair. Do you agree?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 16]
```diff
@@ -38,27 +40,35 @@ public EnricherImpl(EntityMergeEngine entityMergeEngine,
...
-      final List<EnrichmentBase> entities = referenceEnrichmentInformation.stream()
-          .map(EnrichmentBaseWrapper::getEnrichmentBase).collect(Collectors.toList());
-      entityMergeEngine.mergeEntities(rdf, entities, referencesForEnrichment);
+    if (valueEnrichmentInformation != null) {
+      final Set<EnrichmentFields> enrichmentFieldsSet = valueEnrichmentInformation.stream()
+          .map(Pair::getSecond).collect(Collectors.toSet());
+      final List<EnrichmentResultBaseWrapper> enrichmentResultBaseWrapperList = valueEnrichmentInformation
+          .stream().map(Pair::getFirst).collect(Collectors.toList());
+
+      if (CollectionUtils.isNotEmpty(enrichmentResultBaseWrapperList)) {
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:63]: You are right, I have changed it

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java (hunk 4 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:65]: It is done

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 17]
```diff
@@ -38,27 +40,35 @@ public EnricherImpl(EntityMergeEngine entityMergeEngine,
...
+        entityMergeEngine
+            .mergeEntities(rdf, enrichmentResultBaseWrapperList, enrichmentFieldsSet);
+      }
+      if (!referenceEnrichmentInformation.isEmpty()) {
+        final List<List<EnrichmentBase>> entities = referenceEnrichmentInformation.stream()
+            .map(EnrichmentResultBaseWrapper::getEnrichmentBaseList).collect(Collectors.toList());
+        entityMergeEngine.mergeEntities(rdf, entities, referencesForEnrichment);
+      }
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:71]: Done

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 18]
```diff
@@ -71,20 +81,30 @@ public void enrichment(RDF rdf) throws EnrichmentException {
   }
 
   @Override
-  public EnrichmentResultList enrichValues(List<InputValue> valuesForEnrichment)
+  //TODO: Input parameter is List<Pair<SearchValue, EnrichmentFields>
+  //TODO: Output is List<Pair<EnrichmentResult, EnrichmentFields>>
+  public List<Pair<EnrichmentResultBaseWrapper, EnrichmentFields>> enrichValues(
+      List<Pair<SearchValue, EnrichmentFields>> valuesForEnrichment)
       throws EnrichmentException {
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:88]: You're right, we should check if it is empty. But returning empty won't cause problems for the methods that use this one?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java (hunk 2 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java:61]: Indeed  that is better, will change it

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java (hunk 5 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:413]: I also think it is a better idea. However, if I call the method like this `mergeEntities(rdf, enrichmentBaseList, null)` it becomes ambiguous

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java (hunk 6 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/utils/EntityMergeEngine.java:426]: Good question, I don't remember why I removed the static, will add it again

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/EnrichmentSearch.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.List;
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+
+/**
+ * JSON helper class for InputValue class
+ *
+ * @author Yorgos.Mamakis@ europeana.eu
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/EnrichmentSearch.java:19]: Oops

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+  public EnrichmentResultBaseWrapper(List<EnrichmentBase> enrichmentBase) {
+    this.enrichmentBase = enrichmentBase;
+  }
+
+  public List<EnrichmentBase> getEnrichmentBase() {
+    return enrichmentBase;
+  }
+
+  /**
+   * Convert a collection of {@link EnrichmentBase} to a list of {@link EnrichmentBaseWrapper}.
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java:47]: I believe this will be needed in one of the unit tests, I have to double check

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
+package eu.europeana.enrichment.utils;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java:1]: I just put where the object `InputValue` was, since they were similar. But I can put it in the same package as EnrichmentReference

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.List;
+import javax.xml.bind.annotation.XmlRootElement;
+
+@XmlRootElement
+@JsonInclude
+public class ReferenceValue {
+
+  private String reference;
+
+  private List<EntityType> entityTypes;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java:13]: You're right, a Set is better, will change that

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+    this.reference = reference;
+  }
+
+  public List<EntityType> getEntityTypes() {
+    return entityTypes;
+  }
+
+  public void setEntityTypes(List<EntityType> entityTypes) {
+    this.entityTypes = entityTypes;
+  }
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/ReferenceValue.java:43]: I have replaced with `new HashSet<>()` when returnin/setting. If there is a better way to do it, let me know!

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
+package eu.europeana.enrichment.utils;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java:1]: Same as above, I just put where the object InputValue was, since they were similar. But I can move it

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+ */
+@XmlRootElement
+@JsonInclude
+public class SearchValue {
+
+  private String value;
+
+  private String language;
+
+  private List<EntityType> entityTypes;
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java:27]: Done

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+    this.language = language;
+  }
+
+  public List<EntityType> getEntityTypes() {
+    return entityTypes;
+  }
+
+  public void setEntityTypes(List<EntityType> entityTypes) {
+    this.entityTypes = entityTypes;
+  }
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/SearchValue.java:67]: Same as above. I have replaced with `new HashSet<>()` when returnin/setting.

File: metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java [hunk 27]
```diff
@@ -42,42 +45,69 @@ public EnrichmentController(EnrichmentService enrichmentService) {
...
   @ApiOperation(value = "Get an enrichment providing a URI that matches an entity about or "
       + "owlSameAs", response = EnrichmentBase.class)
   @ResponseBody
   @ApiResponses(value = {@ApiResponse(code = 400, message = "Error processing the result")})
-  public EnrichmentBase enrichByEntityAboutOrOwlSameAs(
+  public EnrichmentBase equivalence(
       @ApiParam("uri") @RequestParam("uri") String uri) {
-    return enrichmentService.enrichByAboutOrOwlSameAs(uri);
+    return enrichmentService
+        .enrichByEquivalenceValues(new ReferenceValue(uri, Collections.EMPTY_LIST));
```
[the remark this answers is not in the document]
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java:85]: Since I changed the type to `Set`, I changed this to `Collections.emptySet()`

File: metis-enrichment/metis-enrichment-rest/src/test/java/eu/europeana/enrichment/rest/OldEnrichmentControllerTest.java (hunk 1 above)
Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-rest/src/test/java/eu/europeana/enrichment/rest/OldEnrichmentControllerTest.java:33]: That's a good point, I will take care of it

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 28]
```diff
@@ -71,20 +76,36 @@ public void enrichment(RDF rdf) throws EnrichmentException {
...
+          () -> enrichmentClient.enrich(searchValues));
+      List<List<EnrichmentBase>> enrichmentBaseLists = result.getEnrichmentBaseResultWrapperList()
+          .stream().map(EnrichmentResultBaseWrapper::getEnrichmentBaseList)
+          .collect(Collectors.toList());
+      List<Pair<EnrichmentBase, EnrichmentFields>> output = new ArrayList<>();
+      for (Pair<SearchValue, EnrichmentFields> pair : valuesForEnrichment) {
+        for (List<EnrichmentBase> enrichmentBases : enrichmentBaseLists) {
+          enrichmentBases.forEach(base -> output.add(new Pair<>(base, pair.getSecond())));
+        }
+      }
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:98]: Hmmm.... I think something goes wrong here. I'm not sure if I looked at this code before (and approved it), but I think this doesn't work. In this implementation you match every input with every output (so if there are n inputs, there will be n outputs and your loop goes through n*n iterations).

Instead you need to only go through n iterations.

Here is how I would do it:

```
      final List<Pair<EnrichmentBase, EnrichmentFields>> output = new ArrayList<>();
      for (int index = 0; index < searchValues.size(); index++){
        final EnrichmentFields fieldType = valuesForEnrichment.get(index).getSecond();
        result.getEnrichmentBaseResultWrapperList().get(index).getEnrichmentBaseList()
                .forEach(base -> output.add(new Pair<>(base, fieldType)));
      }
      return output;
```

Does this make sense?

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 29]
```diff
@@ -4,23 +4,25 @@
...
-import eu.europeana.enrichment.utils.InputValue;
+import eu.europeana.enrichment.api.external.SearchValue;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.stream.Collectors;
 import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.math3.util.Pair;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:21]: I think it is more common to use `org.apache.commons.lang3.tuple.Pair` instead, but I don't think it matters. What does matter is consistency: whatever you use elsewhere you should use here.

Comment (JoanaCMS) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:21]: I agree with you, consistency matters and we should use the same everywhere.
I didn't know that `org.apache.commons.lang3.tuple.Pair` is more common, I will change it
